ddtrace/contrib/cassandra/session.py

```diff
@@ -31,32 +36,133 @@ def traced_connect(func, instance, args, kwargs):
     session = func(*args, **kwargs)
     if not isinstance(session.execute, wrapt.FunctionWrapper):
         # FIXME[matt] this should probably be private.
-        setattr(session, 'execute', wrapt.FunctionWrapper(session.execute, traced_execute))
+        setattr(session, 'execute_async', wrapt.FunctionWrapper(session.execute_async, traced_execute_async))
     return session
 
-def traced_execute(func, instance, args, kwargs):
-    cluster = getattr(instance, 'cluster', None)
+def _close_span_on_success(result, future):
+    span = getattr(future, CURRENT_SPAN, None)
+    if not span:
+        log.debug('traced_set_final_result was not able to get the current span from the ResponseFuture')
+        return
+    with span:
+        span.set_tags(_extract_result_metas(cassandra.cluster.ResultSet(future, result)))
+
+def traced_set_final_result(func, instance, args, kwargs):
+    result = kwargs.get('result') or args[0]
+    _close_span_on_success(result, instance)
+    return func(*args, **kwargs)
+
+def _close_span_on_error(exc, future):
+    span = getattr(future, CURRENT_SPAN, None)
+    if not span:
+        log.debug('traced_set_final_exception was not able to get the current span from the ResponseFuture')
+        return
+    with span:
+        # FIXME how should we handle an exception that hasn't be rethrown yet
+        try:
+            raise exc
+        except:
+            span.set_exc_info(*sys.exc_info())
```

Comment:
in that case it's better not to use a `try-except` because we're overriding the stacktrace at this point (other than introducing an overhead). Looking at their code, it seems that everything that goes through `_set_final_exception()` has already been caught by another component, so in any case we don't have a stacktrace. What we can do instead, is handling the `Exception` object manually, with something like:
```python
try:
  # handling the exception manually because we 
  # don't have an ongoing exception here
  span.error = 1
  span.set_tag(errors.ERROR_MSG, exc.args[0])
  span.set_tag(errors.ERROR_TYPE, exc.__class__.__name__)
except Exception:
  log.debug('Unable to se error')  # provide a better message
finally:
  span.finish()
```

Two things about the implementation above:
* we should make the implementation "safe" so we can use a `try-except-finally` statement, or simply make all accessors safe. The try is really efficient unless there is an exception, that should never happen in general (unless `exc` is not always an `Exception`)
* be sure to check the span tags in our tests